File: src/test/java/uk/co/reecedunn/intellij/plugin/marklogic/tests/api/ItemTest.java
/*
 * Copyright (C) 2017 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.marklogic.tests.api;

import junit.framework.TestCase;
import uk.co.reecedunn.intellij.plugin.marklogic.api.Item;
import uk.co.reecedunn.intellij.plugin.marklogic.tests.Query;

import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class ItemTest extends TestCase {
    // region Lexical Types

    @Query("()")
    public void testEmptySequence() {
        Item xcc = Item.Companion.create("()", "empty-sequence()");
        assertThat(xcc.getContent(), is("()"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("empty-sequence()"));

        Item rest = Item.Companion.create("()", "text/plain", "empty-sequence()");
        assertThat(rest.getContent(), is("()"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("empty-sequence()"));
    }

    @Query("1.5")
    public void testXsDecimal() {
        Item xcc = Item.Companion.create("1.5", "xs:decimal");
        assertThat(xcc.getContent(), is("1.5"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:decimal"));

        Item rest = Item.Companion.create("1.5", "text/plain", "decimal");
        assertThat(rest.getContent(), is("1.5"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("decimal"));
    }

    @Query("1e5")
    public void testXsDouble() {
        Item xcc = Item.Companion.create("1e5", "xs:double");
        assertThat(xcc.getContent(), is("1e5"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:double"));

        Item rest = Item.Companion.create("1e5", "text/plain", "double");
        assertThat(rest.getContent(), is("1e5"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("double"));
    }

    @Query("15")
    public void testXsInteger() {
        Item xcc = Item.Companion.create("15", "xs:integer");
        assertThat(xcc.getContent(), is("15"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:integer"));

        Item rest = Item.Companion.create("15", "text/plain", "integer");
        assertThat(rest.getContent(), is("15"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("integer"));
    }

    @Query("'abc'")
    public void testXsString() {
        Item xcc = Item.Companion.create("abc", "xs:string");
        assertThat(xcc.getContent(), is("abc"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:string"));

        Item rest = Item.Companion.create("abc", "text/plain", "string");
        assertThat(rest.getContent(), is("abc"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("string"));
    }

    // endregion
    // region Node Construction

    @Query("array-node { 1, 2 }")
    public void testArrayNode() {
        Item xcc = Item.Companion.create("[1, 2]", "array-node()");
        assertThat(xcc.getContent(), is("[1, 2]"));
        assertThat(xcc.getContentType(), is("application/json"));
        assertThat(xcc.getItemType(), is("array-node()"));

        Item rest = Item.Companion.create("[1, 2]", "application/json", "array-node()");
        assertThat(rest.getContent(), is("[1, 2]"));
        assertThat(rest.getContentType(), is("application/json"));
        assertThat(rest.getItemType(), is("array-node()"));
    }

    @Query("<a b='c'/>/@*")
    public void testAttribute() {
        Item xcc = Item.Companion.create("c", "attribute()");
        assertThat(xcc.getContent(), is("c"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("attribute()"));

        Item rest = Item.Companion.create("c", "text/plain", "attribute()");
        assertThat(rest.getContent(), is("c"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("attribute()"));
    }

    @Query("binary { 'ab' }")
    public void testBinary() {
        Item xcc = Item.Companion.create("«", "binary()");
        assertThat(xcc.getContent(), is("«"));
        assertThat(xcc.getContentType(), is("application/octet-stream"));
        assertThat(xcc.getItemType(), is("binary()"));

        Item rest = Item.Companion.create("«", "application/x-unknown-content-type", "binary()");
        assertThat(rest.getContent(), is("«"));
        assertThat(rest.getContentType(), is("application/octet-stream"));
        assertThat(rest.getItemType(), is("binary()"));
    }

    @Query("boolean-node { true() }")
    public void testBooleanNode() {
        Item xcc = Item.Companion.create("true", "boolean-node()");
        assertThat(xcc.getContent(), is("true"));
        assertThat(xcc.getContentType(), is("application/json"));
        assertThat(xcc.getItemType(), is("boolean-node()"));

        Item rest = Item.Companion.create("true", "application/json", "boolean-node()");
        assertThat(rest.getContent(), is("true"));
        assertThat(rest.getContentType(), is("application/json"));
        assertThat(rest.getItemType(), is("boolean-node()"));
    }

    @Query("<!-- a -->")
    public void testComment() {
        Item xcc = Item.Companion.create("<!-- a -->", "comment()");
        assertThat(xcc.getContent(), is("<!-- a -->"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("comment()"));

        Item rest = Item.Companion.create("<!-- a -->", "text/plain", "comment()");
        assertThat(rest.getContent(), is("<!-- a -->"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("comment()"));
    }

    @Query("document { <a/> }")
    public void testDocument() {
        Item xcc = Item.Companion.create("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<a/>", "document-node()");
        assertThat(xcc.getContent(), is("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<a/>"));
        assertThat(xcc.getContentType(), is("application/xml"));
        assertThat(xcc.getItemType(), is("document-node()"));

        Item rest = Item.Companion.create("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<a/>", "application/xml", "document-node()");
        assertThat(rest.getContent(), is("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<a/>"));
        assertThat(rest.getContentType(), is("application/xml"));
        assertThat(rest.getItemType(), is("document-node()"));
    }

    @Query("<a/>")
    public void testElement() {
        Item xcc = Item.Companion.create("<a/>", "element()");
        assertThat(xcc.getContent(), is("<a/>"));
        assertThat(xcc.getContentType(), is("application/xml"));
        assertThat(xcc.getItemType(), is("element()"));

        Item rest = Item.Companion.create("<a/>", "application/xml", "element()");
        assertThat(rest.getContent(), is("<a/>"));
        assertThat(rest.getContentType(), is("application/xml"));
        assertThat(rest.getItemType(), is("element()"));
    }

    @Query("null-node {}")
    public void testNullNode() {
        Item xcc = Item.Companion.create("null", "null-node()");
        assertThat(xcc.getContent(), is("null"));
        assertThat(xcc.getContentType(), is("application/json"));
        assertThat(xcc.getItemType(), is("null-node()"));

        Item rest = Item.Companion.create("null", "application/json", "null-node()");
        assertThat(rest.getContent(), is("null"));
        assertThat(rest.getContentType(), is("application/json"));
        assertThat(rest.getItemType(), is("null-node()"));
    }

    @Query("number-node { 2 }")
    public void testNumberNode() {
        Item xcc = Item.Companion.create("2", "number-node()");
        assertThat(xcc.getContent(), is("2"));
        assertThat(xcc.getContentType(), is("application/json"));
        assertThat(xcc.getItemType(), is("number-node()"));

        Item rest = Item.Companion.create("2", "application/json", "number-node()");
        assertThat(rest.getContent(), is("2"));
        assertThat(rest.getContentType(), is("application/json"));
        assertThat(rest.getItemType(), is("number-node()"));
    }

    @Query("object-node { 1: 2 }")
    public void testObjectNode() {
        Item xcc = Item.Companion.create("{\"1\": 2}", "object-node()");
        assertThat(xcc.getContent(), is("{\"1\": 2}"));
        assertThat(xcc.getContentType(), is("application/json"));
        assertThat(xcc.getItemType(), is("object-node()"));

        Item rest = Item.Companion.create("{\"1\": 2}", "application/json", "object-node()");
        assertThat(rest.getContent(), is("{\"1\": 2}"));
        assertThat(rest.getContentType(), is("application/json"));
        assertThat(rest.getItemType(), is("object-node()"));
    }

    @Query("<?a?>")
    public void testProcessingInstruction() {
        Item xcc = Item.Companion.create("<?a?>", "processing-instruction()");
        assertThat(xcc.getContent(), is("<?a?>"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("processing-instruction()"));

        Item rest = Item.Companion.create("<?a?>", "text/plain", "processing-instruction()");
        assertThat(rest.getContent(), is("<?a?>"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("processing-instruction()"));
    }

    @Query("<a>abc</a>/text()")
    public void testText() {
        Item xcc = Item.Companion.create("abc", "text()");
        assertThat(xcc.getContent(), is("abc"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("text()"));

        Item rest = Item.Companion.create("abc", "text/plain", "text()");
        assertThat(rest.getContent(), is("abc"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("text()"));
    }

    // endregion
    // region Function Return Types

    @Query("true()")
    public void testXsBoolean() {
        Item xcc = Item.Companion.create("true", "xs:boolean");
        assertThat(xcc.getContent(), is("true"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:boolean"));

        Item rest = Item.Companion.create("true", "text/plain", "boolean");
        assertThat(rest.getContent(), is("true"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("boolean"));
    }

    @Query("map:map()")
    public void testMapMap() {
        // NOTE: Although this is JSON data, the REST API does not give enough
        // information to identify it as JSON, and the XCC API does not identify
        // it as a map:map.

        Item xcc = Item.Companion.create("{}", "json:object");
        assertThat(xcc.getContent(), is("{}"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("json:object"));

        Item rest = Item.Companion.create("{}", "text/plain", "map");
        assertThat(rest.getContent(), is("{}"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("map"));
    }

    // endregion
    // region Type Constructors

    @Query("xs:float(1.5)")
    public void testXsFloat() {
        Item xcc = Item.Companion.create("1.5", "xs:float");
        assertThat(xcc.getContent(), is("1.5"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:float"));

        Item rest = Item.Companion.create("1.5", "text/plain", "float");
        assertThat(rest.getContent(), is("1.5"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("float"));
    }

    @Query("xs:yearMonthDuration('P2Y')")
    public void testXsYearMonthDuration() {
        Item xcc = Item.Companion.create("P2Y", "xs:yearMonthDuration");
        assertThat(xcc.getContent(), is("P2Y"));
        assertThat(xcc.getContentType(), is("text/plain"));
        assertThat(xcc.getItemType(), is("xs:yearMonthDuration"));

        Item rest = Item.Companion.create("P2Y", "text/plain", "yearMonthDuration");
        assertThat(rest.getContent(), is("P2Y"));
        assertThat(rest.getContentType(), is("text/plain"));
        assertThat(rest.getItemType(), is("yearMonthDuration"));
    }

    // endregion
}

File: src/test/java/uk/co/reecedunn/intellij/plugin/marklogic/tests/api/ItemTest.kt
/*
 * Copyright (C) 2017 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.marklogic.tests.api

import junit.framework.TestCase
import uk.co.reecedunn.intellij.plugin.marklogic.api.Item
import uk.co.reecedunn.intellij.plugin.marklogic.tests.Query

import org.hamcrest.CoreMatchers.*
import org.hamcrest.MatcherAssert.assertThat
import uk.co.reecedunn.intellij.plugin.marklogic.api.primitiveToItemType
import java.math.BigDecimal
import java.math.BigInteger

class ItemTest : TestCase() {
    // region primitiveToItemType :: node types

    fun testPrimitiveToItemType_NodeTypes() {
        assertThat(primitiveToItemType("array-node()"), `is`("array-node()"))
        assertThat(primitiveToItemType("attribute()"), `is`("attribute()"))
        assertThat(primitiveToItemType("binary()"), `is`("binary()"))
        assertThat(primitiveToItemType("boolean-node()"), `is`("boolean-node()"))
        assertThat(primitiveToItemType("comment()"), `is`("comment()"))
        assertThat(primitiveToItemType("document-node()"), `is`("document-node()"))
        assertThat(primitiveToItemType("element()"), `is`("element()"))
        assertThat(primitiveToItemType("node()"), `is`("node()"))
        assertThat(primitiveToItemType("null-node()"), `is`("null-node()"))
        assertThat(primitiveToItemType("number-node()"), `is`("number-node()"))
        assertThat(primitiveToItemType("object-node()"), `is`("object-node()"))
        assertThat(primitiveToItemType("processing-instruction()"), `is`("processing-instruction()"))
        assertThat(primitiveToItemType("text()"), `is`("text()"))
    }

    // endregion
    // region primitiveToItemType :: MarkLogic types

    fun testPrimitiveToItemType_MarkLogicTypes() {
        assertThat(primitiveToItemType("array"), `is`("json:array"))
        assertThat(primitiveToItemType("box"), `is`("cts:box"))
        assertThat(primitiveToItemType("circle"), `is`("cts:circle"))
        assertThat(primitiveToItemType("complex-polygon"), `is`("cts:complex-polygon"))
        assertThat(primitiveToItemType("linestring"), `is`("cts:linestring"))
        assertThat(primitiveToItemType("map"), `is`("map:map"))
        assertThat(primitiveToItemType("object"), `is`("json:object"))
        assertThat(primitiveToItemType("period"), `is`("cts:period"))
        assertThat(primitiveToItemType("point"), `is`("cts:point"))
        assertThat(primitiveToItemType("polygon"), `is`("cts:polygon"))
        assertThat(primitiveToItemType("region"), `is`("cts:region"))

        // CTS order types
        assertThat(primitiveToItemType("confidence-order"), `is`("cts:confidence-order"))
        assertThat(primitiveToItemType("document-order"), `is`("cts:document-order"))
        assertThat(primitiveToItemType("fitness-order"), `is`("cts:fitness-order"))
        assertThat(primitiveToItemType("index-order"), `is`("cts:index-order"))
        assertThat(primitiveToItemType("quality-order"), `is`("cts:quality-order"))
        assertThat(primitiveToItemType("score-order"), `is`("cts:score-order"))
        assertThat(primitiveToItemType("unordered"), `is`("cts:unordered"))

        // CTS query types (not a full list)
        assertThat(primitiveToItemType("collection-query"), `is`("cts:collection-query"))
        assertThat(primitiveToItemType("element-query"), `is`("cts:element-query"))
        assertThat(primitiveToItemType("field-range-query"), `is`("cts:field-range-query"))

        // CTS reference types (not a full list)
        assertThat(primitiveToItemType("collection-reference"), `is`("cts:collection-reference"))
        assertThat(primitiveToItemType("element-reference"), `is`("cts:element-reference"))
        assertThat(primitiveToItemType("field-reference"), `is`("cts:field-reference"))
    }

    // endregion
    // region primitiveToItemType :: XMLSchema

    fun testPrimitiveToItemType_XMLSchema() {
        assertThat(primitiveToItemType("anyURI"), `is`("xs:anyURI"))
        assertThat(primitiveToItemType("base64Binary"), `is`("xs:base64Binary"))
        assertThat(primitiveToItemType("boolean"), `is`("xs:boolean"))
        assertThat(primitiveToItemType("date"), `is`("xs:date"))
        assertThat(primitiveToItemType("dateTime"), `is`("xs:dateTime"))
        assertThat(primitiveToItemType("dayTimeDuration"), `is`("xs:dayTimeDuration"))
        assertThat(primitiveToItemType("decimal"), `is`("xs:decimal"))
        assertThat(primitiveToItemType("double"), `is`("xs:double"))
        assertThat(primitiveToItemType("duration"), `is`("xs:duration"))
        assertThat(primitiveToItemType("float"), `is`("xs:float"))
        assertThat(primitiveToItemType("gDay"), `is`("xs:gDay"))
        assertThat(primitiveToItemType("gMonth"), `is`("xs:gMonth"))
        assertThat(primitiveToItemType("gMonthDay"), `is`("xs:gMonthDay"))
        assertThat(primitiveToItemType("gYear"), `is`("xs:gYear"))
        assertThat(primitiveToItemType("gYearMonth"), `is`("xs:gYearMonth"))
        assertThat(primitiveToItemType("hexBinary"), `is`("xs:hexBinary"))
        assertThat(primitiveToItemType("integer"), `is`("xs:integer"))
        assertThat(primitiveToItemType("QName"), `is`("xs:QName"))
        assertThat(primitiveToItemType("string"), `is`("xs:string"))
        assertThat(primitiveToItemType("time"), `is`("xs:time"))
        assertThat(primitiveToItemType("untypedAtomic"), `is`("xs:untypedAtomic"))
        assertThat(primitiveToItemType("yearMonthDuration"), `is`("xs:yearMonthDuration"))
    }

    // endregion
    // region Item.create :: empty-sequence()

    @Query("()")
    fun testEmptySequence() {
        val item = Item.create("()", "empty-sequence()")
        assertThat(item.content, `is`("()"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("empty-sequence()"))
        assertThat(item.toString(), `is`("()"))
    }

    // endregion
    // region Item.fromType :: BigDecimal

    fun testFromBigDecimal() {
        val item = Item.fromType(BigDecimal.valueOf(15, 1))
        assertThat(item.content, `is`("1.5"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:decimal"))
        assertThat(item.toString(), `is`("1.5"))
    }

    // endregion
    // region Item.fromType :: BigInteger

    fun testFromBigInteger() {
        val item = Item.fromType(BigInteger.valueOf(15))
        assertThat(item.content, `is`("15"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:integer"))
        assertThat(item.toString(), `is`("15"))
    }

    // endregion
    // region Item.fromType :: Boolean

    fun testFromBoolean() {
        val item = Item.fromType(true)
        assertThat(item.content, `is`("true"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:boolean"))
        assertThat(item.toString(), `is`("true"))
    }

    // endregion
    // region Item.fromType :: Byte

    fun testFromByte() {
        val item = Item.fromType(20.toByte())
        assertThat(item.content, `is`("20"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:byte"))
        assertThat(item.toString(), `is`("20"))
    }

    // endregion
    // region Item.fromType :: Double

    fun testFromDouble() {
        val item = Item.fromType(1.5)
        assertThat(item.content, `is`("1.5"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:double"))
        assertThat(item.toString(), `is`("1.5"))
    }

    // endregion
    // region Item.fromType :: Float

    fun testFromFloat() {
        val item = Item.fromType(1.5f)
        assertThat(item.content, `is`("1.5"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:float"))
        assertThat(item.toString(), `is`("1.5"))
    }

    // endregion
    // region Item.fromType :: Integer

    fun testFromInteger() {
        val item = Item.fromType(15)
        assertThat(item.content, `is`("15"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:int"))
        assertThat(item.toString(), `is`("15"))
    }

    // endregion
    // region Item.fromType :: Long

    fun testFromLong() {
        val item = Item.fromType(15.toLong())
        assertThat(item.content, `is`("15"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:long"))
        assertThat(item.toString(), `is`("15"))
    }

    // endregion
    // region Item.fromType :: Short

    fun testFromShort() {
        val item = Item.fromType(15.toShort())
        assertThat(item.content, `is`("15"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:short"))
        assertThat(item.toString(), `is`("15"))
    }

    // endregion
    // region Item.fromType :: String

    fun testFromString() {
        val item = Item.fromType("abc")
        assertThat(item.content, `is`("abc"))
        assertThat<String>(item.contentType, `is`("text/plain"))
        assertThat(item.itemType, `is`("xs:string"))
        assertThat(item.toString(), `is`("abc"))
    }

    // endregion
    // region Item.withMimeType

    fun testWithMimeType() {
        val item = Item.withMimeType("<>", "text/turtle")
        assertThat(item.content, `is`("<>"))
        assertThat<String>(item.contentType, `is`("text/turtle"))
        assertThat(item.itemType, `is`("xs:string"))
        assertThat(item.toString(), `is`("<>"))
    }

    // endregion
}
